sourcetracker/_sourcetracker.py

```diff
@@ -9,195 +9,203 @@
-        entry is the sum of the counts of features j in all samples which were
-        considered part of source i.
+    sample_metadata, feature_table : pd.DataFrame, pd.DataFrame
+        Input tables with unshared samples removed and ordered equivalently.
 
-    Notes
-    -----
-    `biom_table` will have S samples = [s0, s1, ... sn]. Each of these samples
-    has a metadata value under `category` = [c0, c1, ... cn]. This function
-    will produce an array with K rows, where the ith row has the data from the
-    ith member of the set of category values. As an example:
-
-    Input table data:
-        s1  s2  s3  s4
-    o1  10  50  10  70
-    o2  0   25  10  5
-    o3  0   25  10  5
-    o4  100 0   10  5
-
-    mf:
-        cat1
-    s1  A
-    s2  A
-    s3  A
-    s4  B
-
-    result of using 'cat1':
-
-    Collapsed table data:
-            o1  o2  o3  o4
-    A       70  35  35  110
-    B       70  5   5   5
-    """
-    envs = []
-    sample_groups = []
-    for s in samples:
-        env = sample_metadata[s][category]
-        try:
-            idx = envs.index(env)
-            sample_groups[idx].append(s)
-        except ValueError:
-            envs.append(env)
-            idx = envs.index(env)
-            sample_groups.append([])
-            sample_groups[idx].append(s)
-
-    collapsed_sources = np.zeros((len(envs), biom_table.shape[0]), dtype=int)
-    for row, group in enumerate(sample_groups):
-        for sample in group:
-            collapsed_sources[row] += biom_table.data(sample, axis='sample',
-                                                      dense=True).astype(int)
-    if sort:
-        indices = np.argsort(envs)
-        return np.array(envs)[indices], collapsed_sources[indices]
+    Raises
+    ------
+    ValueError
+        If no shared samples are found.
+    '''
+    ss = np.intersect1d(sample_metadata.index, feature_table.index)
+    if ss.size == 0:
+        raise ValueError('There are no shared samples between the feature '
+                         'table and the sample metadata. Ensure that you have '
+                         'passed the correct files.')
+    elif ss.size == sample_metadata.shape[0] == feature_table.shape[0]:
+        _sm = sample_metadata.copy()
+        _ft = feature_table.copy()
     else:
-        return np.array(envs), collapsed_sources
+        _sm = sample_metadata.loc[np.in1d(sample_metadata.index, ss), :].copy()
+        _ft = feature_table.loc[np.in1d(feature_table.index, ss), :].copy()
+    return _sm, _ft.loc[_sm.index, :]
 
 
-def subsample_sources_sinks(sources_data, sinks, feature_table, sources_depth,
-                            sinks_depth):
-    '''Rarify data for sources and sinks.
+def get_samples(sample_metadata, col, value):
+    '''Return samples which have `value` under `col`.'''
+    return sample_metadata.index[sample_metadata[col] == value].copy()
+
+
+def collapse_source_data(sample_metadata, feature_table, source_samples,
```

Comment:
It does seem to make sense to generalize the name of this at least - ``collapse_data`` for the function name, and ``samples`` instead of ``source_samples``, since there is nothing source-specific about this.